src/main/java/au/org/ala/biocache/dao/SearchDAOImpl.java

```diff
@@ -911,120 +1054,118 @@ public void writeCoordinatesToStream(SearchRequestParams searchParams,OutputStre
-                                        }
+                                        // throttle the download by sleeping
+                                        Thread.sleep(localThrottle);
                                     }
                                     qr = runSolrQuery(splitByFacetQuery, null, downloadBatchSize, startIndex, "_docid_", "asc");
                                 } else {
                                     qr = null;
                                 }
                             }
-                            return recordsForThread;
+                            return recordsForThread.get();
                         }
                     };
-                    futures.add(pool.submit(solrCallable));
+                    solrCallables.add(solrCallable);
                 }
     
-                //check the futures until all have finished
-                int totalDownload = 0;
-                Set<Future<Integer>> completeFutures = new HashSet<Future<Integer>>();
-                boolean allComplete = false;
-                while(!allComplete){
-                    for(Future future: futures){
-                        if(!completeFutures.contains(future)){
-                            if(future.isDone()){
-                                totalDownload += (Integer) future.get();
-                                completeFutures.add(future);
-                            }
-                        }
-                    }
-                    allComplete = completeFutures.size() == futures.size();
-                    if(!allComplete){
-                        Thread.sleep(1000);
+                // TODO: Do we support downloads running for longer than 1 day??
+                List<Future<Integer>> futures = nextExecutor.invokeAll(solrCallables, 1, TimeUnit.DAYS);
+                AtomicInteger totalDownload = new AtomicInteger(0);
+                for(Future<Integer> future: futures){
+                    if(future.isDone()){
+                        totalDownload.addAndGet(future.get());
                     }
                 }
-                pool.shutdown();
-                out.flush();
 
                 long finish = System.currentTimeMillis();
                 long timeTakenInSecs = (finish-start)/1000;
                 if(timeTakenInSecs <= 0) timeTakenInSecs = 1;
-                logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                if(logger.isInfoEnabled()) {
+                    logger.info("Download of " + resultsCount + " records in " + timeTakenInSecs + " seconds. Record/sec: " + resultsCount.intValue()/timeTakenInSecs);
+                }
             } finally {
-                rw.finalise();
+                try {
+                    concurrentWrapper.finalise();
+                } finally {
+                    try {
+                        writerThread.join(100000);
```

Comment:
Futures will not complete if they are blocked on adding elements to the queue, but the consequence of that is that they will be using one of the threads in the fixed thread pool while blocking. In some ways that is the end goal, and we can increase the number of threads in the fixed thread pool using solr.downloadquery.maxthreads to prevent monopolisation without having the number of threads spike too high.

It is difficult to guess what that parameter should be right now, as there are throughput enhancements by avoiding the two synchronisations that were occurring inside of writeResultsFromIndexToStream that were blocking/context-switching a large number of threads on the other facet result threads. There are also throughput enhancements by avoiding a large number of string creations for logging when it is switched off that could also change the responsiveness. I will make it configurable and we can tune it at runtime.